README.md
# OTMql4AMQP

### OTMql4AMQP - AMQP bindings for MQL4 Python
https://github.com/OpenTrading/OTMql4AMQP/

This project allows the OpenTrading Metatrader-Python bridge
(https://github.com/OpenTrading/OTMql4Py/)
to work with RabbitMQ, or probably any version of AMQP,
the emerging standard for high performance enterprise messaging,
for asynchronous communications between financial and trading applications.
It builds on OTMql4Py, and requires it as a pre-requisite.

In your Python, you must have installed Pika:
https://pypi.python.org/pypi/pika/
Pika offers the advantage of being pure Python: no DLLs to compile
or install. Pika communicates with AMQP servers, the most common
open-source one being RabbitMQ http://www.rabbitmq.com
You will need an AMPQ server installed, configured and running
to use this project. The code assumes the default server setup
of username: guest, password: guest, virtual host: /.

**This is a work in progress - a developers' pre-release version.**

It works on builds > 6xx, but the documentation of the wiring up
of an asnychronous event loop under Python still needs to be done,
as well as more tests and testing on different versions.
Only Python 2.7.x is supported.

The project wiki should be open for editing by anyone logged into GitHub:
**Please report any system it works or doesn't work on in the wiki:
include the Metatrader build number, the origin of the metatrader exe,
the Windows version, and the AMQP server version and version of the Pika.**
This code in known to run under Linux Wine (1.7.x), so this project
bridges Metatrader to RabbitMQ under Linux.

### Installation

For the moment there is no installer: just "git clone" or download the
zip from github.com and unzip into an empty directory. Then recursively copy
the folder MQL4 over the MQL4 folder of your Metatrader installation. It will
not overwrite any Mt4 system files; if it overwrites any OTMql4Py files,
the files are in fact identical (e.g. __init__.py).

### Project

Please file any bugs in the issue tracker:
https://github.com/OpenTrading/OTMql4AMQP/issues

Use the Wiki to start topics for discussion:
https://github.com/OpenTrading/OTMql4AMQP/wiki
It's better to use the wiki for knowledge capture, and then we can pull
the important pages back into the documentation in the share/doc directory.
You will need to be signed into github.com to see or edit in the wiki.

You may want to also look at:
https://github.com/OpenTrading/OpenTrader
which provides a command-line interpreter to speak to an OTMql4AMQP
enabled Metatrader.

### Testing

If you open a command window and go to your `MQL4\Python` directory
run this command first
`
python OTMql427/PikaListener.py "#"
`
Then open another command window and go to your `MQL4\Python` directory
run this command second
`
python OTMql427/PikaChart.py "foo"
`
You should soon see `['foo']` appear in the first window.

If it does not work, then solve this problem first, because it
won't work from Python under Mt4 if it doesn't work here.

If you leave the first command running, you should see the tick
and bar information broadcast by the calls in `OnTimer` and `OnTick` in
`MQL4/Experts/OTMql4/OTPyTestPikaEA.mq4`

`OnTimer` is called every iTIMER_INTERVAL_SEC (10 sec.)
which allows us to use Python to look for Pika inbound messages,
or execute a stack of calls from Python to us in Metatrader.


CHANGELOG.txt
commit 69fd8a6f261253ded4dd6658d4aa887f06f66928
Date:   Thu Feb 25 15:06:56 2016 +0200

    Added OTPyAMQPTest.mq4

commit 11667c7031d2af601b9c0ddc39e6df9b1768cb7d
Date:   Tue Feb 23 05:15:54 2016 +0200

    Updated documentation

commit 8cd878ec634143bc45538644b606b7a96f585b2d
Date:   Mon Feb 22 21:57:28 2016 +0200

    Updated documentation

commit 886d833d7fa1b486cbc89a6f419b3e8e0c335645
Date:   Mon Feb 22 21:44:45 2016 +0200

    Updated documentation

commit 407204b2e4a050372ecd92de77fef5f7cb3c5f86
Date:   Mon Feb 22 16:18:25 2016 +0200

    Updated documentation

commit 2ae100b7c0ffa8171d5637a608e030be14287134
Date:   Sat Feb 20 07:38:58 2016 +0200

    Updated documentation

commit 0b24b7cbb10e9deea213f8b3bdca17213b25bb07
Date:   Sat Aug 22 07:03:08 2015 +0300

    Github creole bugs workaround

commit 1cc63c1d0a9df09ad937edb44133e1484dc0164b
Date:   Sat Aug 22 07:00:31 2015 +0300

    Github creole bugs workaround

commit e5aff8a5c6c6acff36259f92decff81c78f388bf
Date:   Sat Aug 22 06:55:09 2015 +0300

    Github creole bugs workaround

commit fdc76062590744982f45b9a5e661f3b08be8fa9d
Date:   Wed Aug 12 21:35:02 2015 +0300

    Fixed using ZeroMQ for OpenTrader

commit 56037a71aaf23359120fb59c20e597d3bfe1bf90
Date:   Mon Jul 13 20:20:24 2015 +0300

     Tried to fix communications for errors but it still needs a reqrite

commit 20b89c2e215ac635beddb6975805ead01a992ca2
Date:   Sun Jul 12 22:29:58 2015 +0300

    Minor refactoring of Mql4Chart

commit 8c240b8f21cd25f30e687f9388aa8a697e57b081
Date:   Wed Jul 1 23:03:49 2015 +0300

    Refactored the threading in OpenTrader

commit 109008feb17cbf9d1a533771987621ce2580cb99
Date:   Fri Jun 19 17:17:58 2015 +0300

    Refactored SMARK For ZmqCmd

commit 39a018c8c60e62e672c6c59f9dbac2a2a8e4823a
Date:   Thu Jun 18 16:57:03 2015 +0300

    Added wiki pages

commit 7617d4244ea1c4c41fd20aa36767d3f771d394e0
Date:   Thu Jun 18 16:40:30 2015 +0300

    Added the wiki.

commit 3bfb015559a897ab7349e649dbc578dee1f5b531
Date:   Tue Jun 16 23:23:32 2015 +0300

    Minor bugfixes.

commit 388ab02d7c2f730dca23cefc1a3b6adbdff2440a
Date:   Tue Jun 16 14:04:02 2015 +0300

    Added some extra checking for Rabbit server errors.

commit 72a2d34f8f647df43b9130ed87284f27d6728efb
Date:   Fri Jun 5 00:07:25 2015 +0300

    Moved share/doc/05PikaCmd2.txt to OpenTrader.

commit 09130207a636ec05d1fa524e367fcfbfa27745ea
Date:   Thu Jun 4 23:43:33 2015 +0300

    Updated documentation.

commit bd955f4daf46473f4e4d85fa6756042786b7b9ae
Date:   Thu Jun 4 16:44:45 2015 +0300

    Moved MQL4/Python/OTMql427/PikaCmd2.py to OpenTrader

commit 2c43ed72bc817662c45ba8d346a8af1240b26842
Date:   Wed Jun 3 09:53:24 2015 +0300

    Extended PikaChart.py and PikaCmd2.py.
    Will probably break PikaCmd2.py out to a new package to support backtesting.

commit e53b51433b28cc4c7dbe28ba7aa3db9d1d9498c8
Date:   Wed May 27 12:09:06 2015 +0300

    Fleshed out PikaCmd2.py and made messaging use more JSON.

commit bb5368a628edb7dbfc719385ab26a33b7a2c3193
Date:   Wed May 27 12:01:56 2015 +0300

    Added the beginning of order handling to PikaCmd2.py

commit b3f56dd04c0fa1e6854f4399d9cb34781dc91d28
Date:   Tue May 19 20:54:30 2015 +0300

    Added MQL4/Python/OTMql427/PikaCmd2.py to communicate withan OTMql4Py
    enabled metatrader. It's a Cmd2 based REPL with some basic commands.

commit 19756b9b2fd5d6c49cd5aadf417c2e8e11e962cd
Date:   Mon May 18 15:41:08 2015 +0300

    Refactored the MQL code into MQL4/Libraries/OTMql4/OTPyChart.mq4
    and started using JSON to communicate to the client.

commit b0ce9f6a42ff0953e126720d318737652ea92633
Date:   Fri May 15 11:29:23 2015 +0300

    Refactoring to pull common code from OTMql4AMQP and OTMql4Zmq into OTMql4Py.

commit cbf013527065a2ed88007fd2918697f5204a76e1
Date:   Fri May 8 19:10:45 2015 +0300

    Minor refactoring.

commit 5e2c45e91f324f172a16a3e5a483d1260fac1de4
Date:   Fri May 8 16:44:56 2015 +0300

    Refactored the Python to make it easier to share with PyZmq.

commit 3509b1988623eb2c8affc391717041102154e760
Date:   Thu May 7 01:11:51 2015 +0300

    Typo bugfix.

commit 5f11449121dc10af0f301c41cfe0613335edf023
Date:   Thu May 7 00:55:08 2015 +0300

    Refactored the Simple message format.

commit a151f7ee6d88e7c42ed3e9397ad0a4b00ff4a752
Date:   Wed May 6 16:02:57 2015 +0300

    Checked in the callme code, even though we are not using it:
    MQL4/Python/OTMql427/PikaCallme.py

commit cb64e6ea234dc51d01cf201c1a3757b4aecfdf80
Date:   Tue May 5 19:41:36 2015 +0300

    Wired up the return channel so that Mt4 can receive commands over RabbitMQ.
    Still need to clean it up and document it, but it works, WITHOUT CRASHES!

commit 0e1141ad799d77055c0beaa429536ceda30489c9
Date:   Sat May 2 14:44:02 2015 +0300

    Renamed MQL4/Include/OTMql4/OTBarInfo.mqh in OTMql4Lib

commit 475e3e23f5d9ea4fe239b2b78403b7c8493a7da4
Date:   Sat May 2 14:29:15 2015 +0300

    Wired in callme but it's not working, and is disabled for now.

commit be212d2f520699bdf9891439de168ae16238813f
Date:   Fri May 1 17:39:26 2015 +0300

    Made an EA that is now working and broadcasting ticks and bars.
    Needs cleaning up with regard to queues and topics, but it's working.
    
    Started playing with callme as a return RPC command channel.
    You'll need to install callme and its prereqs first.
    Could be a whole lot of fun...

commit 4544dd12070da1ed45297c01801c83337441f002
Date:   Tue Apr 28 16:45:44 2015 +0300

    First Python code checkin.

commit 3da983b2bb3d11ea0e2981b4cff6fc224f91d526
Date:   Tue Apr 28 16:16:17 2015 +0300

    Initial commit.

commit 0382dfcb69deefd09caa530d782a14b5dfba2034
Date:   Tue Apr 28 13:07:24 2015 +0000

    Initial commit


MQL4/Python/OTMql427/PikaChart.py
# -*-mode: python; py-indent-offset: 4; indent-tabs-mode: nil; encoding: utf-8-dos; coding: utf-8 -*-

"""
A PikaChart object is a simple abstraction to encapsulate a Mt4 chart
that has a RabbitMQ connection on it. There should be only one connection for
the whole application, so it is set as the module variable oCONNECTION.

This module can be run from the command line to test RabbitMQ with a listener
such as OTMql427/PikaListener.py. Give the message you want to publish
as arguments to this script, or --help to see the options.
"""

import sys, logging
import time
import traceback
import pika

oLOG = logging

from Mq4Chart import Mq4Chart, oFindChartByName, lFindAllCharts
from PikaListener import PikaMixin
from Mt4SafeEval import sPySafeEval
from SimpleFormat import lUnFormatMessage

if True:
    ePikaCallme = "PikaCallme disabled "
    PikaCallme = None
else:
    # The callme server is optional and may not be installed.
    # But it might be a whole lot of fun it it works.
    # It has prerequisities: kombu httplib2 amqp
    try:
        import PikaCallme
        ePikaCallme = ""
    except ImportError as e:
        ePikaCallme = "Failed to import PikaCallme: " + str(e)
        PikaCallme = None

class PikaChart(Mq4Chart, PikaMixin):

    iDeliveryMode = 1 # (non-persisted)
    sContentType = 'text/plain'

    def __init__(self, sChartId, **dParams):
        Mq4Chart.__init__(self, sChartId, **dParams)
        PikaMixin.__init__(self, sChartId, **dParams)
        self.sChartId = sChartId
        self.sQueueName = "listen-for-commands"

    def vPikaRecvOnListener(self, sQueueName, lBindingKeys):
        if self.oListenerChannel is None:
            self.eBindBlockingListener(sQueueName, lBindingKeys)
        assert self.oListenerChannel, "vPikaRecvOnListener: oListenerChannel is null"
        #FixMe: does this block?
        # http://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume
        # no-wait no-wait
        # not in pika.channel.Channel.basic_consume
        self.oListenerChannel.basic_consume(self.vPikaCallbackOnListener,
                                            queue=self.oListenerQueueName,
                                            exclusive=True,
                                            no_ack=False
        )

    def eHeartBeat(self, iTimeout=0):
        """
        The heartbeat is usually called from the Mt4 OnTimer.
        We push a simple Print exec command onto the queue of things
        for Mt4 to do if there's nothing else happening. This way we get
        a message in the Mt4 Log,  but with a string made in Python.
        """
        if self.oQueue.empty():
            # only push if there is nothing to do
            sTopic = 'exec'
            sMark = "%15.5f" % time.time()
            sInfo = "PY: " +sMark
            sMess = "%s|%s|0|%s|Print|%s" % (sTopic, self.sChartId, sMark, sInfo,)
            self.eMq4PushQueue(sMess)

        # while we are here flush stdout so we can read the log file
        # whilst the program is running
        sys.stdout.flush()
        sys.stderr.flush()

        # now for the hard part - see if there is anything to receive
        # does this block? do we set a timeout?
        if self.oListenerChannel is None:
            # , 'json.#'
            lBindingKeys = ['cmd.#', 'eval.#']
            self.vPikaRecvOnListener(self.sQueueName, lBindingKeys)

        # This is the disabled callme server code
        if iTimeout > 0 and self.oListenerServer:
            # join it and do a little work but dont block for long
            # cant use self.oListenerServer.wait()
            print "DEBUG: listening on server"
            self.oListenerServer.drain_event(iTimeout=iTimeout)

        return ""

    def vPikaDispatchOnListener(self, sBody, oProperties=None):
        #? do we need the oProperties for content-type?
        # 'content_encoding', 'content_type', 'correlation_id', 'decode', 'delivery_mode', 'encode', 'expiration', 'headers', 'message_id', 'priority', 'reply_to', 'timestamp', 'type', 'user_id'

        lArgs = lUnFormatMessage(sBody)
        sMsgType = lArgs[0]
        sChart = lArgs[1]
        sIgnore = lArgs[2] # should be a hash on the payload
        sMark = lArgs[3]
        sVerbMaybe = lArgs[4]
        gPayload = lArgs[4:] # overwritten below

        if sMsgType == 'cmd':
            # FixMe; dispatch to the right chart
            lChartInstances = lFindAllCharts()
            if not lChartInstances:
                # this should never happen
                sys.stdout.write("ERROR: vPikaDispatchOnListener no charts\n")
                self.eMq4PushQueue(sBody)
                return
            if sChart.find('ANY') >= 0:
                #? use self?
                oElt =lChartInstances[0]
                oElt.eMq4PushQueue(sBody)
                return
            if sChart.find('ALL') >= 0:
                for oElt in lChartObjects:
                    oElt.eMq4PushQueue(sBody)
                return

            o = oFindChartByName(sChart)
            if o is not None:
                o.eMq4PushQueue(sBody)
                return

            sys.stdout.write("WARN: vPikaDispatchOnListener unrecognized sBody " +sBody +"\n")
            sys.stdout.flush()

            self.eMq4PushQueue(sBody)
            return

        #? assume eval is on any chart?
        if sMsgType == 'eval':
            # unused
            lRetval = ['retval']
            lRetval += lArgs[1:3]
            sCmd = lArgs[4]
            if len(lArgs) > 5:
                sCmd += '(' +lArgs[5] +')'
            sRetval = sPySafeEval(sCmd)
            if sRetval.find('ERROR:') >= 0:
                lRetval += ['error', sRetval]
            else:
                lRetval += ['string', sRetval]
            sRetval = '|'.join(lRetval)
            # FixMe; dispatch to the right chart
            self.eReturnOnSpeaker('retval', sRetval, sBody)
            return

    def vPikaCallbackOnListener(self, oChannel, oMethod, oProperties, sBody):
        assert sBody, "vPikaCallbackOnListener: no sBody received"
        oChannel.basic_ack(delivery_tag=oMethod.delivery_tag)
        sMess = "vPikaCallbackOnListener Listened: %r" % sBody
        sys.stdout.write("INFO: " +sMess +"\n")
        sys.stdout.flush()
        # we will assume that the sBody
        # is a "|" seperated list of command and arguments
        # FixMe: the sMess must be in the right format
        # FixMe: refactor for multiple charts:
        # we must push to the right chart
        try:
            self.vPikaDispatchOnListener(sBody, oProperties)
        except Exception as e:
            sys.stdout.write("ERROR: " +str(e) +"\n" + \
                             traceback.format_exc(10) +"\n")
            sys.stdout.flush()
            sys.exc_clear()

    # unused
    def eStartCallmeServer(self, sId='Mt4Server'):
        # The callme server is optional and may not be installed
        if not PikaCallme:
            return ePikaCallme
        if self.oListenerServer is None:
            oServer = PikaCallme.Server(server_id=sId)
            # danger - we are running this in the main thread
            # self.oListenerThread = _run_server_thread(oServer)
            oServer.connect()
            oServer.register_function(sPySafeEval, 'sPySafeEval')
            oServer.register_function(self.eMq4PushQueue, 'eMq4PushQueue')
            self.oListenerServer = oServer
            print "DEBUG: started the callme server %d" % id(oServer)

        return ""


def iMain():
    from PikaArguments import oParseOptions

    sUsage = __doc__.strip()
    oArgParser = oParseOptions(sUsage)
    oArgParser.add_argument('lArgs', action="store",
                            nargs="*",
                            help="the message to send (required)")
    oOptions = oArgParser.parse_args()
    lArgs = oOptions.lArgs

    assert lArgs, "Give the command you want to send as arguments to this script"

    sSymbol = 'ANY'
    iPeriod = 0
    sTopic = 'cmd'
    sMark = "%15.5f" % time.time()
    sMsg = "%s|%s|0|%s|%s|str|%s" % (sTopic, sSymbol+str(iPeriod), sMark, '|'.join(lArgs),)

    oChart = None
    try:
        oChart = PikaChart('oANY_0_FFFF_0', **oOptions.__dict__)
        iMax = 1
        i = 0
        print "Sending: %s %d times " % (sMsg, iMax,)
        while i < iMax:
            # send a burst of iMax copies
            oChart.eSendOnSpeaker('cmd', sMsg)
            i += 1
        # print "Waiting for message queues to flush..."
        time.sleep(1.0)
    except KeyboardInterrupt:
        pass
    except Exception as e:
        print(traceback.format_exc(10))

    try:
        if oChart:
            print "DEBUG: Waiting for message queues to flush..."
            oChart.bCloseConnectionSockets(oOptions)
            time.sleep(1.0)
    except KeyboardInterrupt:
        pass

if __name__ == '__main__':
    iMain()


MQL4/Python/OTMql427/PikaArguments.py
# -*-mode: python; py-indent-offset: 4; indent-tabs-mode: nil; encoding: utf-8-dos; coding: utf-8 -*-

"""
This file declares oParseOptions in a separate file so that the
arguments parsing can be uniform between applications that use it.

"""

from argparse import ArgumentParser

def oParseOptions(sUsage):
    """
    Look at the bottom of PikaListener.py and PikaChart.py for iMain
    functions that use the oParseOptions that is returned here.
    This function returns an ArgumentParser instance, so that you
    can override it before you call it to parse_args.
    """
    oArgParser = ArgumentParser(description=sUsage)
    # rabbit
    oArgParser.add_argument("-a", "--address",
                            action="store",
                            dest="sHostAddress",
                            default="127.0.0.1",
                            help="the TCP address to subscribe on (default 127.0.0.1)")
    oArgParser.add_argument("-o", "--pubport", action="store",
                            dest="iPubPort", type=int, default=5672,
                            help="the TCP port number to publish to (default 5672)")
    oArgParser.add_argument("-u", "--username", action="store",
                            dest="sUsername", default="guest",
                            help="the username for the connection (default guest)")
    oArgParser.add_argument("-p", "--password", action="store",
                            dest="sPassword", default="guest",
                            help="the password for the connection (default guest)")
    oArgParser.add_argument("-e", "--exchange", action="store",
                            dest="sExchangeName", default="Mt4",
                            help="sExchangeName for the connection (default Mt4)")
    oArgParser.add_argument("-P", "--mt4dir", action="store",
                            dest="sMt4Dir", default="",
                            help="directory for the installed Metatrader")
    oArgParser.add_argument("-i", "--virtual", action="store",
                            dest="sVirtualHost", default="/",
                            help="the VirtualHost for the connection (default /)")
    oArgParser.add_argument("-q", "--queue", action="store",
                            dest="sQueueName", default="listen-for-ticks",
                            help="the VirtualHost for the connection (default listen-for-ticks)")
    oArgParser.add_argument("-v", "--verbose", action="store",
                            dest="iDebugLevel", type=int, default=4,
                            help="the verbosity, 0 for silent 4 max (default 4)")
    return oArgParser



MQL4/Python/OTMql427/__init__.py
# -*-mode: python; py-indent-offset: 4; indent-tabs-mode: nil; encoding: utf-8-dos; coding: utf-8 -*-

import sys, os, thread, threading, traceback

# import rpdb2; rpdb2.start_embedded_debugger('foobar')

# There are/were(2.4) some atexit handlers that were crashing Mt4
# at exit because thay were killing the process, so Mt4 got killed
# when you unloaded the py27.dll. This may not be needed anymore...
import atexit
atexit._exithandlers = atexit._exithandlers[:-1]

# There are/were(2.4) some logging handlers that were polluting Mt4
# at shutdown
import logging
oLOG = logging
def vShutupShutdown(*args, **dArgs):
    pass
oLOG.shutdown = vShutupShutdown

def sPySafeEval(sPyCode):
    """
    This wraps its string argument in a try:/except:
    so that it always succeeds. If there was an error,
    then the string returned starts with ERROR:
    followed by a description of the error.

    In the caller you should have something like:

      if (StringFind(uRetval, "ERROR:", 0) == 0) {
        Print("Error in Python evaluating: " + uSource + "\n" + res);
        <do something as a result of the failure>
      }

    In fact, you should probably do ALL your calls into Python
    using sPySafeEval unless you know what you are calling
    traps errors, and calls sys.exc_clear() if there is an error.

    """

    dGlobals = sys.modules['__main__'].__dict__
    s = "try:\n    sRetval=" + sPyCode + "\nexcept Exception,e:\n    sRetval='ERROR: '+str(e)"
    try:
        k = compile(s, '<string>', 'exec')
    except Exception, e:
        sRetval = "ERROR: Python error compiling " + sPyCode+ ': '+str(e)
        sys.stderr.write(sRetval+'\n')
        traceback.print_exc(None, sys.stderr)
        sys.exc_clear()
        return sRetval

    try:
        eval(k, dGlobals, dGlobals)
        if dGlobals['sRetval']:
            sRetval = str(dGlobals['sRetval'])
        else:
            sRetval = ""
    except Exception, e:
        sRetval = "ERROR: Python error evaling " + sPyCode+ ': '+str(e)
        sys.stderr.write(sRetval+'\n')
        traceback.print_exc(None, sys.stderr)
        sys.exc_clear()
        return sRetval

    if sRetval.find('ERROR:') == 0:
        sys.exc_clear()
    return sRetval


try:
    import win32ui, win32con

    def iMessageBox(sMsg, sTitle, iType, iIcon):
        """
        A frivolity that demonstrates that Mark Hammond's
        win32 code is all callable by Python under Mt4.
        """
        i = win32ui.MessageBox(sMsg, sTitle, iType | iIcon)
        # while != 0 ?
        win32ui.PumpWaitingMessages()
        return i
except ImportError:
    def iMessageBox(sMsg, sTitle, iType, iIcon):
        return -1

# from win32con
MB_OK = 0
MB_OKCANCEL = 1
MB_ABORTRETRYIGNORE = 2
MB_YESNOCANCEL = 3
MB_YESNO = 4
MB_RETRYCANCEL = 5
MB_ICONHAND = 16
MB_ICONQUESTION = 32
MB_ICONEXCLAMATION = 48
MB_ICONASTERISK = 64
MB_ICONWARNING = MB_ICONEXCLAMATION
MB_ICONERROR = MB_ICONHAND
MB_ICONINFORMATION = MB_ICONASTERISK
MB_ICONSTOP = MB_ICONHAND
IDOK = 1
IDCANCEL = 2
IDABORT = 3
IDRETRY = 4
IDIGNORE = 5
IDYES = 6
IDNO = 7
IDCLOSE = 8
IDHELP = 9

lLOG_ARRAY = ["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG", "TRACE"]

def vLog(iLevel, sMsg, *lArgs, **dKeys):
    """
    Level                                Numeric value
    ------                               -----
    CRITICAL                             50
    ERROR                                40
    WARNING                              30
    INFO                                 20
    DEBUG                                10
    NOTSET                               0

From OFLibLogging.mqh:
#define LOG_PANIC 0 // unused
#define LOG_ERROR 1
#define LOG_WARN 2
#define LOG_INFO 3
#define LOG_DEBUG 4
#define LOG_TRACE 5
#define LOG_MAX 5
    """
    iId = thread.get_ident()
    try:
        assert 0 <= int(iLevel) <= 5
    except:
        iLevel = 4
    try:
        # oLOG.log(50 - iLevel*4, sMsg)
        if iLevel <= 0:
            # oLOG.critical(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLOG_ARRAY[iLevel], sMsg
        elif iLevel <= 1:
            # oLOG.error(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLOG_ARRAY[iLevel], sMsg
        elif iLevel <= 2:
            # oLOG.warning(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLOG_ARRAY[iLevel], sMsg
        elif iLevel <= 3:
            # oLOG.info(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLOG_ARRAY[iLevel], sMsg
        else:
            # oLOG.debug(sMsg, *lArgs, **dKeys)
            print str(iId) + " " +lLOG_ARRAY[4], sMsg
        sys.stdout.flush()
        # ValueError when logging to a closed file
    except IOError:
        # [Errno 28] No space left on device - is that IOError?
        # probably filled the disk
        pass
    except Exception, e:
        print str(iId) + " ERROR: in vLog ", str(e), iLevel, sMsg
        sys.exc_clear()

oTKINTER_ROOT = None
sSTDOUT_FD = None

def bStartTkinter():
    """
    Strange but true: you can run Tkinter on Python under Mt4!
    Two event loops, two GUIs; what more could you want.

    We are not planning to take this any further.
    """

    global oTKINTER_ROOT, sSTDOUT_FD
    import Tkinter
    if oTKINTER_ROOT is None:
        if not hasattr(sys, 'argv'):
            # __file__
            sys.argv = ['py27.py']
        try:
            # should start this in a thread and leave it running?
            oTKINTER_ROOT = Tkinter.Tk()
            oTKINTER_ROOT.withdraw()
        except Exception, e:
            sShowError('bStartTkinter', "Error starting Tkinter\n%s" % (e,))
            return False

        if True:
            sShowInfo('bStartTkinter', "Started Tkinter")
        else:
            # always pegs to CPU at max
            try:
                import TkFileIO
                # _default_root
                oTop = oTKINTER_ROOT
                frame = TkFileIO.MainFrame(oTop, title='Python IO')
                oTop.deiconify()
                oTop.tkraise()
                # i = oTop.tk.dooneevent()
                oTop.update()
                sys.stdout = frame.shell
                sys.stderr = frame.shell
                print "INFO: bStartTkinter: Started TkFileIO"
                oTop.update()
            except Exception, e:
                sShowError('bStartTkinter', "Error starting TkFileIO\n%s" % (e,))
                # return False

        return True

    elif oTKINTER_ROOT.__class__ == Tkinter.Tk:
        sShowInfo('bStartTkinter', "Already Started Tkinter")
        return True
    else:
        sShowError('bStartTkinter', "Error starting Tkinter - %s" % (
            oTKINTER_ROOT.__class__,))
        return False

def sShowError(sTitle, sMsg):
    global oTKINTER_ROOT
    if oTKINTER_ROOT:
        import tkMessageBox
        tkMessageBox.showerror(sTitle, sMsg)
    elif sSTDOUT_FD:
        sSTDOUT_FD.write('ERROR : ' + sTitle + ' ' + sMsg + '\n')

def sShowInfo(sTitle, sMsg):
    global oTKINTER_ROOT, sSTDOUT_FD
    if oTKINTER_ROOT is not None:
        import tkMessageBox
        tkMessageBox.showinfo(sTitle, sMsg)
    elif sSTDOUT_FD is not None:
        sSTDOUT_FD.write('INFO : ' + sTitle + ' ' + sMsg + '\n')

def eStartFile(sStdout):
    # idempotent
    global sSTDOUT_FD

    if sStdout == "":
        # setting sStdout = "" means dont log to file
        return ""

    if sSTDOUT_FD is None:
        try:
            if os.path.isabs(sStdout):
                # absolute filename
                pass
            elif sStdout.find('..') >= 0 or sStdout.find('\\') >= 0:
                # relative pathname to this directory e.g. ..\..\Foo
                sStdout = os.path.join(os.path.dirname(__file__), sStdout)
            else:
                # stick in ..\..\Logs
                sStdout = os.path.join(
                    os.path.dirname(os.path.dirname(os.path.dirname(__file__))),
                    'Logs', sStdout)
            # dont delete it - it may be a pipe
            sSTDOUT_FD = open(sStdout, 'w', 1)
            sys.stdout = sys.stderr = sSTDOUT_FD
            assert sys.stdout != sys.__stdout__
            # we should always see our version in the log
            print sys.version
            sys.stdout.flush()
            assert os.path.isfile(sStdout)
        except Exception, e:
            # may be in trouble logging here if stdout was not opened
            sys.exc_clear()
            return str(e)

        try:
            # oLOG.basicConfig(level = logging.DEBUG)
            print "vPyInit - Opened %s %s" % (sSTDOUT_FD.name, sStdout,)
            print "vPyInit - Thread " + threading.currentThread().getName() + \
                  " number " + str(thread.get_ident())
            return ""
        except Exception, e:
            sMsg = "vPyInit - Error opening %s\n%s" % (sStdout, str(e),)
            sys.exc_clear()
            print sMsg
            return sMsg

    elif sSTDOUT_FD:
        # sSTDOUT_FD.seek(0, 0)
        # sSTDOUT_FD.truncate(0)
        return ""

def ePyInit(sStdout):
    return eStartFile(sStdout)

def vPyInit(sStdout):
    eStartFile(sStdout)
    # bStartTkinter()

def vPyDeInit():
    global oTKINTER_ROOT, sSTDOUT_FD
    if sSTDOUT_FD:
        try:
            sys.stdout = sys.__stdout__
            sys.stderr = sys.__stderr__
            sName = sSTDOUT_FD.name
            # sShowInfo('vPyDeInit', "Closing %s" % (sName,))
            sSTDOUT_FD.write('INFO : vPyDeInit ' + "Closing outfile %s\n" % (sName,))
            # oLOG.shutdown()
            sSTDOUT_FD.flush()
            sSTDOUT_FD.close()
            sSTDOUT_FD = None
        except Exception, e:
            # You probably have not stdout so no point in logging it!
            print "Error closing %s\n%s" % (sSTDOUT_FD, str(e),)
            sys.exc_clear()

    if oTKINTER_ROOT:
        oTKINTER_ROOT.destroy()
        oTKINTER_ROOT = None

    sys.exc_clear()

def test():
    # changing sys.stdout can't be done under doctest
    if os.path.isfile('test.txt'): os.remove('test.txt')

    vPyInit('test.txt')
    assert os.path.isfile('test.txt')

    vPyInit('test.txt')
    assert os.path.isfile('test.txt')
    assert sys.stdout != sys.__stdout__
    assert sys.stderr != sys.__stderr__

    # This can't be done under doctest
    s = sPySafeEval('foobar')
    assert s.find('ERROR:') == 0
    assert s == "ERROR: name 'foobar' is not defined"

    vLog(0, "Level 0")
    vLog(1, "Level 1")
    vLog(2, "Level 2")
    vLog(3, "Level 3")
    vLog(4, "Level 4")

    # oLOG.root.setLevel(40)
    # oLOG.debug("NONONO oLOG.debug")
    # vLog(4, "NONO vLog 4")

    vPyDeInit()
    assert sys.stdout == sys.__stdout__
    assert sys.stderr == sys.__stderr__
    assert oTKINTER_ROOT is None
    assert sSTDOUT_FD is None

    # should check contents of test.txt
    oFd = open('test.txt', 'r')
    sContents = oFd.read()
    oFd.close()
    assert sContents.find("Level 4") > 0
    assert sContents.find("NONO") < 0

def vTestTkinter():
    if False:
        bStartTkinter()
        assert oTKINTER_ROOT
        sShowError('Foo', 'Bar')

def vTestMessageBox():
    i = iMessageBox("Hi there", "Yes No Cancel",
                    MB_YESNOCANCEL, MB_ICONINFORMATION)
    if i == IDCANCEL:
        pass
    elif i == IDYES:
        pass
    elif i == IDOK:
        pass
    # win32ui.PumpWaitingMessages()

